### PR TITLE
mindforger: remove use of deprecated QtWebKit

### PR DESCRIPTION
**Pull request by cyplo**
Removed references to `QtWebKit` and migrated to `QtWebEngine`. See https://github.com/NixOS/nixpkgs/issues/53079 for context

Hit some minor snags where the compilation with `QtWebEngine` did not work out of the box, just a few missed references to `QtWebFrame` - fixed that with a patch. As a side note - I will talk to the developer to try to get the patches upstreamed.

###### Motivation for this change
I've noticed that `mindforger` no longer installs from master because of build failures coming from `QtWebKit`, took this opportunity to remove references to it as it's deprecated.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 122476` at 80286e3d run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>mindforger</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>error: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-flags-spaces.md">no-flags-spaces</a>

`qmakeFlags` cannot contain spaces, please use `qmakeFlagsArray` in Bash.

Near pkgs/applications/editors/mindforger/default.nix:32:3:
```
   |
32 |   qmakeFlags = [ "-r mindforger.pro" "CONFIG+=mfnoccache" "CONFIG+=mfwebengine"] ;
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/applications/editors/mindforger/default.nix:25:29:
```
   |
25 |   patches = [ ./paths.patch ./webengine.patch ] ;
   |                             ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/applications/editors/mindforger/default.nix:25:15:
```
   |
25 |   patches = [ ./paths.patch ./webengine.patch ] ;
   |               ^
```
</li>
  </ul>
</details>

File: pkgs/applications/editors/mindforger/default.nix [hunk 1]
```diff
@@ -18,18 +18,18 @@ mkDerivation rec {
   };
 
   nativeBuildInputs = [ qmake wrapGAppsHook ] ;
-  buildInputs = [ qtbase qtwebkit cmark-gfm ] ;
+  buildInputs = [ qtbase qtwebengine cmark-gfm  ] ;
 
   doCheck = true;
 
-  patches = [ ./build.patch ] ;
+  patches = [ ./paths.patch ./webengine.patch ] ;
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/mindforger/default.nix:25]:
```suggestion
  patches = [
    ./paths.patch
    ./webengine.patch
  ] ;
```

File: pkgs/applications/editors/mindforger/default.nix [hunk 2]
```diff
@@ -18,18 +18,18 @@ mkDerivation rec {
...
 
   postPatch = ''
     substituteInPlace lib/src/install/installer.cpp --replace /usr "$out"
     substituteInPlace app/resources/gnome-shell/mindforger.desktop --replace /usr "$out"
   '';
 
-  qmakeFlags = [ "-r mindforger.pro" "CONFIG+=mfnoccache" ] ;
+  qmakeFlags = [ "-r mindforger.pro" "CONFIG+=mfnoccache" "CONFIG+=mfwebengine"] ;
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/mindforger/default.nix:32]:
```suggestion
  qmakeFlags = [
    "-r"
    "mindforger.pro"
    "CONFIG+=mfnoccache"
    "CONFIG+=mfwebengine"
  ];
```

**Comment by veprbl** [comment]
@cyplo Would you mind submitting your nice patch upstream?

**Comment by cyplo** [comment]
@SuperSandro2000 - not sure if splitting the arrays into multiple lines makes it more readable, is this more idiomatic ? Also - not sure if having `-r` on a separate line is equivalent to having it in one string ? 

@veprbl - yeah defo, once this lands in nixos I will submit upstream - ideally with someone giving this a go on a MacOS box before, as I don't have access to any.

**Comment by veprbl** [comment]
@cyplo The preferred way is to first have the patch upstream and then pull it down using `fetchpatch` in the expression. I agree with your response to the formatting comments.

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 - not sure if splitting the arrays into multiple lines makes it more readable, is this more idiomatic ? Also - not sure if having `-r` on a separate line is equivalent to having it in one string ? 
> 
> @veprbl - yeah defo, once this lands in nixos I will submit upstream - ideally with someone giving this a go on a MacOS box before, as I don't have access to any.

It is less likely to cause merge conflicts and beginning with a line length of ~120 it should always be done to keep the lines readable on the GitHub web interface.

Also please upstream patches beforehand because the developer can give important feedback which we can't know about.

**Comment by veprbl** [comment]
Fixing darwin build would additionally require
```diff
diff --git a/pkgs/applications/editors/mindforger/default.nix b/pkgs/applications/editors/mindforger/default.nix
index 512e0f12441..bbbdd34a9fc 100644
--- a/pkgs/applications/editors/mindforger/default.nix
+++ b/pkgs/applications/editors/mindforger/default.nix
@@ -27,6 +27,9 @@ mkDerivation rec {
   postPatch = ''
     substituteInPlace lib/src/install/installer.cpp --replace /usr "$out"
     substituteInPlace app/resources/gnome-shell/mindforger.desktop --replace /usr "$out"
+    for f in app/app.pro lib/lib.pro; do
+      substituteInPlace "$f" --replace "QMAKE_CXX = g++" ""
+    done
   '';
 
   qmakeFlags = [ "-r mindforger.pro" "CONFIG+=mfnoccache" ] ;
```

**Comment by cyplo** [comment]
nice one, thank you both ! will try to contact upstream and let's see how it goes :)

File: pkgs/applications/editors/mindforger/default.nix (hunk 1 above)
Comment (veprbl) [inline pkgs/applications/editors/mindforger/default.nix:25]:
```suggestion
  patches = [
    ./paths.patch
    (fetchpatch {
      url = "https://github.com/dvorka/mindforger/commit/9e4febbecb56c54d52909966712920560116f736.diff";
      sha256 = "sha256-HYKRGVDiWC+7TyxaO3FdVXcPfaNFgqn04RvDRwvaJkY=";
    })
  ] ;
```

**Comment by cyplo** [comment]
the upstream did not respond - will apply suggestions here and reference the patch commit and let you know :)

**Comment by cyplo** [comment]
@veprbl @SuperSandro2000 if you could review that would be grand
OfBorg seems stuck on the darwin build btw, not sure if related to the PR

**Review by veprbl — APPROVED** [review]
Works on darwin

**Comment by cyplo** [comment]
awesome work, thank you !

**Pull request merged**